File: ios/Arculus/CSDK/CSDKAPICall.swift
import CoreNFC
import CSDK

enum CSDKError: Int32, Error, LocalizedError {
    case nullPointer = -100
    case nullAppleObj = -101
    case nullCalloc = -102
    case wrongResponseLength = -103
    case wrongResponseData = -104
    case wrongStatusWord = -105
    case wrongDataLength = -106
    case wrongParamLength = -107
    case wrongPIN = -108
    case invalidParam = -109
    case encryptionNotInit = -110
    case extOrChainNotSupported = -111
    case apiChainNotSupported = -112
    case unknownError = -113
    case apduExceedsChainLength = -114
    case extApduSupportRequired = -115
    case apduTooBig = -116
    case walletNotSelected = -117
    
    var errorDescription: String? {
        switch self {
        case .nullPointer: return "Null pointer encountered"
        case .nullAppleObj: return "Wallet session object is NULL"
        case .nullCalloc: return "Unable to allocate memory"
        case .wrongResponseLength: return "Card response length is incorrect/unexpected"
        case .wrongResponseData: return "Card response not valid"
        case .wrongStatusWord: return "Card response status not expected"
        case .wrongDataLength: return "Data length of payload is invalid"
        case .wrongParamLength: return "Parameter size validation failed"
        case .wrongPIN: return "Wrong PIN"
        case .invalidParam: return "Invalid Parameter"
        case .encryptionNotInit: return "NFC Session encryption was not initialized"
        case .extOrChainNotSupported: return "Card doesn't support extended APDUs or chaining"
        case .apiChainNotSupported: return "API is deprecated and requires Chaining"
        case .unknownError: return "An unknown error has occurred"
        case .apduExceedsChainLength: return "APDU too big to do chaining"
        case .extApduSupportRequired: return "Extended APDU not supported but required"
        case .apduTooBig: return "APDU too big"
        case .walletNotSelected: return "Wallet not selected"
        }
    }
    
    init(rc: Int32) {
        self = CSDKError(rawValue: rc) ?? CSDKError.unknownError
    }
}

class CSDKAPICallError: Error, LocalizedError {
    enum CSDKAPICallErrorType {
        case apduCreationFailed
        case requestCreationFailed(String)
        case responseParsingFailed(String)
        case walletResponseError(Int32)
    }
    
    var callErrorType: CSDKAPICallErrorType?
    
    init(_ errorType: CSDKAPICallErrorType) {
        self.callErrorType = errorType
    }
    
    init() {
        self.callErrorType = nil
    }
    
    var errorDescription: String? {
        switch callErrorType {
        case .apduCreationFailed:
            return "Failed to create APDU from data"
        case let .requestCreationFailed(command):
            return "Failed to create \(command) request"
        case let .responseParsingFailed(command):
            return "Failed to parse \(command) response"
        case let .walletResponseError(rc):
            var len: size_t = 0
            var message: UnsafePointer<CChar>?
            
            let response = WalletErrorMessage(rc, &message, &len)
            
            guard response == CSDK_OK, let message = message else {
                return CSDKError(rc: rc).localizedDescription
            }
            
            let description = String(cString: message)
            
            guard description != "Unknown error code" else {
                return CSDKError(rc: rc).localizedDescription
            }
            
            return description
        default: return "Unknown error"
        }
    }
    
    static func validateWalletResponseCode(rc: Int32) throws {
        guard rc == CSDK_OK else {
            throw CSDKAPICallError(.walletResponseError(rc))
        }
    }
}


class CSDKAPICall<T>: CSDKAPICommand<T> {
    final override func execute() async throws -> ResponseType {
        fatalError("execute() must not be called - use execute(tag:) instead")
    }
    
    func request() async throws -> [Data] {
        fatalError("request must be overridden")
    }
    
    func response(bytes: [UInt8]) async throws -> ResponseType {
        fatalError("response must be overridden")
    }
    
    func sendCommand(_ data: Data?, tag: NFCISO7816Tag) async throws -> [UInt8] {
        guard let data = data else {
            preconditionFailure("cannot send empty request")
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            guard let apdu = NFCISO7816APDU(data: data) else {
                return continuation.resume(throwing: CSDKAPICallError(.apduCreationFailed))
            }
            
            tag.sendCommand(apdu: apdu) { response, sw1, sw2, error in
                if let error = error {
                    return continuation.resume(throwing: error)
                }
                
                let data = response + Data([sw1, sw2])
                
                continuation.resume(returning: Array(data))
            }
        }
    }
    
    func execute(tag: NFCISO7816Tag) async throws -> ResponseType {
        let data = try await request()
        
        assert(data.count == 1, "request should return exactly one data item in the array")
        
        let bytes = try await sendCommand(data.first, tag: tag)
        
        return try await response(bytes: bytes)
    }
}

File: android/src/main/java/com/aptoslabs/arculus/csdk/CSDKAPICall.kt
package com.aptoslabs.arculus.csdk

import android.nfc.tech.IsoDep
import com.sun.jna.Pointer
import com.sun.jna.ptr.PointerByReference

class CSDKAPICallRequestCreationException : Exception("Failed to create request")
class CSDKAPICallResponseParsingException : Exception("Failed to parse response")

class CSDKError(rc: Int) : Exception(getErrorMessage(rc)) {
  companion object {
    private const val CSDK_ERR_NULL_POINTER = -100
    private const val CSDK_ERR_NULL_APPLETOBJ = -101
    private const val CSDK_ERR_NULL_CALLOC = -102
    private const val CSDK_ERR_WRONG_RESPONSE_LENGTH = -103
    private const val CSDK_ERR_WRONG_RESPONSE_DATA = -104
    private const val CSDK_ERR_WRONG_STATUS_WORD = -105
    private const val CSDK_ERR_WRONG_DATA_LENGTH = -106
    private const val CSDK_ERR_WRONG_PARAM_LENGTH = -107
    private const val CSDK_ERR_WRONG_PIN = -108
    private const val CSDK_ERR_INVALID_PARAM = -109
    private const val CSDK_ERR_ENCRYPTION_NOT_INIT = -110
    private const val CSDK_ERR_EXT_OR_CHAIN_NOT_SUPORTED = -111
    private const val CSDK_ERR_API_CHAIN_NOT_SUPORTED = -112
    private const val CSDK_ERR_UNKNOWN_ERROR = -113
    private const val CSDK_ERR_APDU_EXCEEDS_CHAIN_LENGTH = -114
    private const val CSDK_ERR_EXTAPDU_SUPPORT_REQUIRED = -115
    private const val CSDK_ERR_APDU_TOO_BIG = -116
    private const val CSDK_ERR_WALLET_NOT_SELECTED = -117

    private fun getErrorMessage(rc: Int): String {
      val message = PointerByReference()
      val len = SizeTByReference()

      if (CSDK.WalletErrorMessage(
          rc,
          message,
          len
        ) == CSDK.CSDK_OK && message.value != Pointer.NULL
      ) {
        val description = message.value.getString(0)

        if (description != "Unknown error code") {
          return description
        }
      }

      return when (rc) {
        CSDK_ERR_NULL_POINTER -> "Null pointer encountered"
        CSDK_ERR_NULL_APPLETOBJ -> "Wallet session object is NULL"
        CSDK_ERR_NULL_CALLOC -> "Unable to allocate memory"
        CSDK_ERR_WRONG_RESPONSE_LENGTH -> "Card response length is incorrect/unexpected"
        CSDK_ERR_WRONG_RESPONSE_DATA -> "Card response not valid"
        CSDK_ERR_WRONG_STATUS_WORD -> "Card response status not expected"
        CSDK_ERR_WRONG_DATA_LENGTH -> "Data length of payload is invalid"
        CSDK_ERR_WRONG_PARAM_LENGTH -> "Parameter size validation failed"
        CSDK_ERR_WRONG_PIN -> "Wrong PIN"
        CSDK_ERR_INVALID_PARAM -> "Invalid Parameter"
        CSDK_ERR_ENCRYPTION_NOT_INIT -> "NFC Session encryption was not initialized"
        CSDK_ERR_EXT_OR_CHAIN_NOT_SUPORTED -> "Card doesn't support extended APDUs or chaining"
        CSDK_ERR_API_CHAIN_NOT_SUPORTED -> "API is deprecated and requires Chaining"
        CSDK_ERR_UNKNOWN_ERROR -> "An unknown error has occurred"
        CSDK_ERR_APDU_EXCEEDS_CHAIN_LENGTH -> "APDU too big to do chaining"
        CSDK_ERR_EXTAPDU_SUPPORT_REQUIRED -> "Extended APDU not supported but required"
        CSDK_ERR_APDU_TOO_BIG -> "APDU too big"
        CSDK_ERR_WALLET_NOT_SELECTED -> "Wallet not selected"
        else -> "Unknown error code: $rc"
      }
    }
  }
}


abstract class CSDKAPICall<ResponseType>(wallet: Pointer) : CSDKAPICommand<ResponseType>(wallet) {
  override suspend fun execute(): ResponseType {
    throw Error("execute() must not be called - use execute(tag:) instead")
  }

  open suspend fun request(): Array<ByteArray> {
    throw NotImplementedError("request must be overridden")
  }

  open suspend fun response(bytes: ByteArray): ResponseType {
    throw NotImplementedError("response must be overridden")
  }

  protected fun sendCommand(data: ByteArray?, tag: IsoDep): ByteArray {
    requireNotNull(data) { "cannot send empty request" }

    return tag.transceive(data)
  }

  open suspend fun execute(tag: IsoDep): ResponseType {
    val data = request()

    require(data.size == 1) { "request should return exactly one data item in the array" }

    val bytes = sendCommand(data.first(), tag)

    return response(bytes)
  }

  protected fun validateWalletResponse(rc: Int) {
    if (rc != CSDK.CSDK_OK) {
      throw CSDKError(rc)
    }
  }
}
